### PR TITLE
ReportTiming callback should record the sendFrameToEngine when it was scheduled

### LINKED ISSUE #144261: `WidgetsBinding.drawFrame` falsely assumes that `sendFramesToEngine` won't turn from true to false
During `WidgetsBinding.drawFrame`, a `firstFrameCallback` is scheduled to be called on the next timing report. The `firstFrameCallback` asserts that `sendFramesToEngine` is true. While this assertion might seem reasonable at a glance, since a timing report is called only after `Rasterizer` renders a frame, which requires `sendFramesToEngine` to be true, it's wrong. The assertion can be triggered if `sendFramesToEngine` turns from true to false between when the frame is sent to the engine and when the timing report callback is invoked.

The case where `sendFramesToEngine` turns false right before `firstFrameCallback` is fired is extremely rare. However, the assumption is wrong anyway. We can assert that the `sendFramesToEngine` value when the callback is registered must be true. We must not assert that it hasn't turned false when the callback is invoked.

### Analysis 

This problem was discovered in an internal test. The test is very simple, something like
```dart
  testWidgets('Culprit test', (tester) async {
    runApp(LocaleApp());
    await tester.pump();
  });
```
where `LocaleApp` is a widget that uses the localization API provided by `MaterialApp`.

When the test is run **on a live device**, it crashes with the following log:
```
'package:flutter/src/widgets/binding.dart': Failed assertion: line 968 pos 16: 'sendFramesToEngine': is not true.
When the exception was thrown, this was the stack:
dart:core-patch/errors_patch.dart 51:61            _AssertionError._doThrowNew
dart:core-patch/errors_patch.dart 40:5             _AssertionError._throwNew
package:flutter/src/widgets/binding.dart 968:16    WidgetsBinding.drawFrame.<fn>
package:flutter/src/scheduler/binding.dart 343:19  SchedulerBinding._executeTimingsCallbacks
dart:ui/hooks.dart 328:13                          _invoke1
dart:ui/platform_dispatcher.dart 611:5             PlatformDispatcher._reportTimings
dart:ui/hooks.dart 278:31                          _reportTimings
```

Deeper analysis shows that this problem is a combination of `testWidgets`, live device testing, `runApp`, and `deferFirstFrame`.

The `sendFramesToEngine` can turn false only if `deferFirstFrame` is called when no frame has been rendered yet. But if no frame has been rendered yet, how can there be a timing reporting callback, which is only invoked when a frame is rendered?

This is because `testWidgets` does some preparation before running a test body: It pumps a dummy widget, then calls `resetFirstFrameSent`.

https://github.com/flutter/flutter/blob/7f2b238c55a191567f6e93a170419aed41d8eb62/packages/flutter_test/lib/src/binding.dart#L1006-L1013

The pumped dummy widget is guaranteed rendered, and leaves a timing reporting callback on hold. This creates an inconsistent state at start of the test body: there's a timing report callback pending (since the UI thread has not `await`ed yet), but the framework thinks that no frame is sent. 

Now, if the test body uses `runApp` (instead of `tester.pump`) so that the widget tree is attached but no frame is rendered, and builds the widget tree, which defers the first frame, and then await, then the timing report callback will be invoked, finding the current value of `sendFramesToEngine` is false!

Apparently, the `testWidgets` leaking the timing reporting callback from the preparation is also a problem, and fixing either `testWidgets` or `WidgetTester` can solve the crash, but the `WidgetTester` should be of higher priority because:
1. `WidgetTester` is a production class
2. The assumption of `WidgetTester` is 100% wrong
3. It might not be trivial to fix: we might not be able to simply waiting for `waitUntilFirstFrameRasterized` since a warm-up frame might never be rendered.

This issue might be related to https://github.com/flutter/flutter/issues/132969

### PR DESCRIPTION
**Pull request by dkwingsmt**
Fixes https://github.com/flutter/flutter/issues/144261

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

File: packages/flutter/test/widgets/binding_live_test.dart [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+// This file is for tests for WidgetsBinding that require a `LiveTestWidgetsFlutterBinding`
+void main() {
+  LiveTestWidgetsFlutterBinding();
+  testWidgets('ReportTiming callback records the sendFramesToEngine when it was scheduled', (WidgetTester tester) async {
+    // This test needs LiveTestWidgetsFlutterBinding for multiple reasons.
+    // First, this was the environment that this bug was discovered.
+    // Second, LiveTestWidgetsFlutterBinding doesn't override
+    // scheduleWarmUpFrame, so that the test can start with no frames rendered
+    // allowing deferFirstFrame to work (see below).
```
Comment (goderbauer) [inline packages/flutter/test/widgets/binding_live_test.dart:16]: I don't understand this paragraph. In my observation, the problem is that the warm-up frame triggered internally by `testWidgets` before the test body executes is actually send to the engine, but after that `resetFirstFrame` is called, which now puts engine and framework in an inconsistent state: The framework now things no frame has been sent (which means `deferFirstFrame` does have an effect) while the engine things the first frame has been sent.

File: packages/flutter/test/widgets/binding_live_test.dart [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    // At the beginning of a `testWidgets`, `TestWidgetsFlutterBinding` pumps a
+    // fixed dummy widget using `runApp`, then `resetFirstFrameSent`. Although
+    // the dummy widget is rendered, its reportTiming callback from the engine
+    // has not arrived yet.
+    //
+    // Note that the `runApp` call also schedules a warm up frame, which would
+    // have directly rendered the frame if the test had used
+    // `AutomatedTestWidgetsFlutterBinding`.
```
Comment (goderbauer) [inline packages/flutter/test/widgets/binding_live_test.dart:25]: This doesn't seem super relevant and adds more confusion to what is actually going on here. Can we just drop this? Also note that "note that" is banned: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-empty-prose

File: packages/flutter/test/widgets/binding_live_test.dart [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    expect(tester.binding.sendFramesToEngine, true);
+    // Push the widget with runApp instead of tester.pump, avoiding rendering a
+    // frame is rendered, which is needed for `deferFirstFrame` later to work.
```
Comment (goderbauer) [inline packages/flutter/test/widgets/binding_live_test.dart:28]: This sentence is busted: "avoiding rendering a frame is rendered".

File: packages/flutter/test/widgets/binding_live_test.dart [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    runApp(const DummyWidget());
+    // Ensure that no widget tree is built and nothing is rendered.
+    expect(find.text('First frame'), findsNothing);
+    // Defer the first frame, making sendFramesToEngine false, and widget tree
+    // will be built but not sent to the engine.
+    tester.binding.deferFirstFrame();
+    // Pump a while, letting the reportTiming callback to run. If it had not
+    // recorded the `sendFramesToEngine` when it was scheduled, it would have
+    // crashed seeing `sendFramesToEngine` being now false!
+    await tester.pump(const Duration(milliseconds: 1));
```
Comment (goderbauer) [inline packages/flutter/test/widgets/binding_live_test.dart:38]: Suggestion:

```suggestion
    expect(tester.binding.sendFramesToEngine, false);
    // Wait for the reportTiming callback (which completes the future below) to run. If the reportTiming callback were to assume that `sendFramesToEngine` is true, the callback would crash.
    await tester.binding.waitUntilFirstFrameRasterized;
```

**Review by goderbauer — COMMENTED** [review]
Overall this seems like the right fix, the assert makes incorrect assumptions since we offer API to reset `sendFramesToEngine`.

File: packages/flutter/test/widgets/binding_live_test.dart (hunk 1 above)
Comment (dkwingsmt) [reply packages/flutter/test/widgets/binding_live_test.dart:16]: Kind of. The `_runTestBody` has:
```dart
    runApp(Container(key: UniqueKey(), child: _preTestMessage)); // Reset the tree to a known state.
    await pump();
```
so a frame is pumped to the engine anyway, warm-up or not. 

And your understanding on the inconsistency is absolutely correct. Now there's a reportTiming callback pending, which is supposed to only happen after rendered frames, while the framework thinks no frames have been rendered, allowing `deferFirstFrame` to take effect. This combined situation of `LiveTestWidgetsFlutterBinding` and `testWidgets` is probably the only case that can trigger this bug.

**Comment by dkwingsmt** [comment]
@goderbauer I've addressed your feedbacks and rewrote some paragraphs. Can you take another look? (I'm filing an issue to record the problem, which I'll link to the test file. )

File: packages/flutter/test/widgets/binding_live_test.dart [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    // Before `testWidgets` executes the test body, it pumps a frame with a
+    // fixed dummy widget, then calls `resetFirstFrameSent`. The pumped frame
+    // schedules a reportTiming call that has yet to arrive.
+    //
+    // This puts the test in an inconsistent state: a reportTiming callback is
+    // supposed to happen only after a frame is rendered, but due to
+    // `resetFirstFrameSent`, the framework thinks no frames have been rendered.
+
+    expect(tester.binding.sendFramesToEngine, true);
+    // Push the widget with runApp instead of tester.pump, avoiding rendering a
```
Comment (goderbauer) [inline packages/flutter/test/widgets/binding_live_test.dart:32]: nitty-nit: For consistency with the other comments put runApp and tester.pump in ` as well?

File: packages/flutter/test/widgets/binding_live_test.dart [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    // frame, which is needed for `deferFirstFrame` later to work.
+    runApp(const DummyWidget());
+    // Verify that no widget tree is built and nothing is rendered.
+    expect(find.text('First frame'), findsNothing);
+    // Defer the first frame, making sendFramesToEngine false, so that widget
```
Comment (goderbauer) [inline packages/flutter/test/widgets/binding_live_test.dart:37]: same with sendFramesToEngine here

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/test/widgets/binding_live_test.dart [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    // tree will be built but not sent to the engine.
+    tester.binding.deferFirstFrame();
+    expect(tester.binding.sendFramesToEngine, false);
+    // Wait for the reportTiming callback (which completes the future below) to
+    // run. If the reportTiming callback were to assume that
+    // `sendFramesToEngine` is true, the callback would crash.
+    await tester.binding.waitUntilFirstFrameRasterized;
+    expect(find.text('First frame'), findsOne);
```
Comment (goderbauer) [inline packages/flutter/test/widgets/binding_live_test.dart:45]: From the test failures: It looks like the test needs a `pump` call before this expect, which makes sense. Sorry to not include that in my previous suggestion.

Comment (dkwingsmt) [reply packages/flutter/test/widgets/binding_live_test.dart:45]: What do you think if I just revert to `    await tester.pump(const Duration(milliseconds: 1));`?

Comment (dkwingsmt) [reply packages/flutter/test/widgets/binding_live_test.dart:45]: I ended up including both.

**Comment by dkwingsmt** [comment]
I also wrote an issue to describe the problem https://github.com/flutter/flutter/issues/144261 and linked it to the test case. This PR will close the issue.

**Pull request merged**